cli/commons/src/Microsoft.Kiota.Cli.Commons/IO/ODataPagingService.cs

```diff
@@ -0,0 +1,147 @@
+                }
+            }
+            catch (JsonException)
+            {
+                // If the response isn't valid JSON, there will be no next link.
+                // TODO: Log warning once logging story is defined
+            }
+        }
+
+        return null;
+    }
+
+    /// <inheritdoc />
+    public override async Task<Stream?> MergePageAsync(Stream? currentResult, PageLinkData newPageData, CancellationToken cancellationToken = default)
+    {
+        if (IsJson(newPageData))
+        {
+            return await MergeJsonStreamsAsync(currentResult, newPageData.Response, newPageData.ItemName, newPageData.NextLinkName, cancellationToken);
+        }
+
+        return null;
+    }
+
+    /// <inheritdoc />
+    public override bool OnBeforeGetPagedData(PageLinkData pageLinkData, bool fetchAllPages = false)
+    {
+        // Set the page size to 999 if the user asked to fetch all pages and top either isn't specified or is invalid
+        if (fetchAllPages && (!pageLinkData.RequestInformation.QueryParameters.TryGetValue("%24top", out var topVal) || (topVal as int?) < 1))
+        {
+            pageLinkData.RequestInformation.QueryParameters["%24top"] = 999;
+        }
+        return true;
+    }
+
+    private bool IsJson(PageLinkData pageLinkData)
+    {
+        return pageLinkData.RequestInformation.Headers.TryGetValue("Accept", out var accepts) && accepts.Contains("json");
+    }
+
+    /// <summary>
+    /// Merges 2 streams of JSON on the property defined by <code>itemName</code>. The property should be a JSON array
+    /// </summary>
+    /// <param name="left">The first stream.</param>
+    /// <param name="right">The second stream.</param>
+    /// <param name="itemName">The name of the array property to merge on.</param>
+    /// <param name="nextLinkName">The name of the property containing the next link name.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    internal async Task<Stream?> MergeJsonStreamsAsync(Stream? left, Stream? right, string itemName = "value", string nextLinkName = "nextLink", CancellationToken cancellationToken = default)
+    {
+        if (left?.CanSeek == true) left?.Seek(0, SeekOrigin.Begin);
+        if (right?.CanSeek == true) right?.Seek(0, SeekOrigin.Begin);
+        if (left == null || right == null)
+        {
+            return left ?? right;
+        }
+
+        JsonNode? nodeLeft = JsonNode.Parse(left);
+        if (left.CanSeek == true) left.Seek(0, SeekOrigin.Begin);
+        JsonNode? nodeRight = JsonNode.Parse(right);
+        if (right.CanSeek == true) right.Seek(0, SeekOrigin.Begin);
+
+        JsonArray? leftArray = null;
+        JsonArray? rightArray = null;
+        if (!string.IsNullOrWhiteSpace(itemName))
+        {
+            if (nodeLeft?[itemName] == null || nodeRight?[itemName] == null)
+            {
+                return left ?? right;
```

Comment:
`left` is never null.

```suggestion
                return left;
```

Reply:
Great catch. The fix is actually different because of the intent. I've made the changes